File: src/leetcode/p701to750/LeetCode739DailyTemperatures.java
package leetcode.p701to750;

import java.util.Stack;

public class LeetCode739DailyTemperatures {
  
  public int[] dailyTemperatures(int[] temperatures) {
    int[] res = new int[temperatures.length];
    Stack<Integer> stack = new Stack<>();
    for (int i = 0; i < temperatures.length; i++) {
      if (stack.isEmpty() || temperatures[stack.peek()] >= temperatures[i]) {
        stack.push(i);
      } else {
        int index = stack.pop();
        res[index] = i - index;
        i--;
      }
    }
    return res;
  }

  public static void main(String[] args) {

  }

}

File: karakoram/kb/kb0700/src/main/kotlin/com/anonymouscorgi/karakoram/kb0700/LeetCode739DailyTemperatures.kt
package com.anonymouscorgi.karakoram.kb0700

import com.anonymouscorgi.karakoram.annotation.Accepted
import com.anonymouscorgi.karakoram.annotation.Medium

@Medium
internal interface LeetCode739DailyTemperatures {

    fun dailyTemperatures(temperatures: IntArray): IntArray

    @Accepted
    object METHOD : LeetCode739DailyTemperatures {

        override fun dailyTemperatures(temperatures: IntArray): IntArray {
            val result = IntArray(temperatures.size)
            val stack = ArrayDeque<Int>()

            for (dayNum in temperatures.indices) {
                while (!stack.isEmpty() && temperatures[stack.last()] < temperatures[dayNum]) {
                    val index = stack.removeLast()
                    result[index] = dayNum - index
                }
                stack.addLast(dayNum)
            }
            return result
        }
    }
}